power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -0,0 +1,471 @@
+                      .elements_per_scenario = elements_per_scenario,
+                      .total_elements = total_elements,
+                      .msg_data = msg_data,
+                      .data = nullptr,
+                      .indptr = {}};
+    }
+
+    bool check_uniform(IdxVector const& counter) {
+        if (batch_size_ < 2) {
+            return true;
+        }
+        return std::transform_reduce(counter.cbegin(), counter.cend() - 1, counter.cbegin() + 1, true,
+                                     std::logical_and{}, std::equal_to{});
+    }
+
+    Idx get_elements_per_scenario(IdxVector const& counter) {
+        bool const is_uniform = check_uniform(counter);
+        if (!is_uniform) {
+            return -1;
+        }
+        if (batch_size_ == 0) {
+            return 0;
+        }
+        return counter.front();
+    }
+
+    void parse_component(Buffer const& buffer) const {
+        component_key_ = buffer.component->name;
+        // handle indptr
+        if (buffer.elements_per_scenario < 0) {
+            // first always zero
+            buffer.indptr.front() = 0;
+            // accumulate sum
+            // TODO Apple Clang cannot compile transform_inclusive_scan correctly
+            // So we disable the good code and write the loop manually
+            // std::transform_inclusive_scan(buffer.msg_data.cbegin(), buffer.msg_data.cend(),
+            //                               buffer.indptr.begin() + 1,
+            //                               std::plus{}, [](auto const& x) { return static_cast<Idx>(x.size()); });
+            for (Idx i = 0; i != batch_size_; ++i) {
+                buffer.indptr[i + 1] = buffer.indptr[i] + static_cast<Idx>(buffer.msg_data[i].size());
+            }
+        }
+        // set nan
+        buffer.component->set_nan(buffer.data, 0, buffer.total_elements);
+        // attributes
+        auto const attributes = [&]() -> std::span<MetaAttribute const* const> {
+            auto const found = attributes_.find(buffer.component->name);
+            if (found == attributes_.cend()) {
+                return {};
+            }
+            return found->second;
+        }();
+        // all scenarios
+        for (scenario_number_ = 0; scenario_number_ != batch_size_; ++scenario_number_) {
+            Idx const scenario_offset = buffer.elements_per_scenario < 0
+                                            ? buffer.indptr[scenario_number_]
+                                            : scenario_number_ * buffer.elements_per_scenario;
+#ifndef NDEBUG
+            if (buffer.elements_per_scenario < 0) {
+                assert(buffer.indptr[scenario_number_ + 1] - buffer.indptr[scenario_number_] ==
+                       static_cast<Idx>(buffer.msg_data[scenario_number_].size()));
+
+            } else {
+                assert(buffer.elements_per_scenario == static_cast<Idx>(buffer.msg_data[scenario_number_].size()));
+            }
+#endif
```

Comment:
you don't need the `#ifndef NDEBUG`. `assert` already implies `#ifndef NDEBUG` and the compiler **should** optimize the empty `if` and `else` blocks away in `Release`/`RelWithDebInfo` mode, and even if it doesn't, a single boolean check is trivial compared to all the other logic happening

```suggestion
            if (buffer.elements_per_scenario < 0) {
                assert(buffer.indptr[scenario_number_ + 1] - buffer.indptr[scenario_number_] ==
                       static_cast<Idx>(buffer.msg_data[scenario_number_].size()));

            } else {
                assert(buffer.elements_per_scenario == static_cast<Idx>(buffer.msg_data[scenario_number_].size()));
            }
```

Reply:
done

Reply:
See
https://github.com/PowerGridModel/power-grid-model/actions/runs/5785220968/job/15677448351?pr=336

Clang-Tidy gives warning in Release build.